PR: https://github.com/odoo/odoo/pull/62891

From: 5b886a747a7d4a91470764321f04b1cc61a9cc73
From: Yannick Tivisse
Date: 2020-12-04 13:38:51

Structural Changes: 6
Total Changes: 16

[FIX] hr: Edit private address fields directly instead of M2O

Purpose
=======

The goal of this commit if double:
- Prevent employees to change the m2o address_home_id, that could lead
  to future inconsistencies on future HR processes
- Allow the employee to modify it own private address, even if the
  access is restricted to administrators for private addresses

closes odoo/odoo#62891

Taskid: 2412387
Signed-off-by: Yannick Tivisse (yti) <yti@odoo.com>

================================= pseudo patch: =================================

--- a/addons/hr/models/res_users.py
+++ b/addons/hr/models/res_users.py
@@ -25,6 +25,14 @@ class User(models.Model):
     employee_parent_id = fields.Many2one(related='employee_id.parent_id', readonly=False, related_sudo=False)
     coach_id = fields.Many2one(related='employee_id.coach_id', readonly=False, related_sudo=False)
     address_home_id = fields.Many2one(related='employee_id.address_home_id', readonly=False, related_sudo=False)
+    private_street = fields.Char(related='address_home_id.street', string="Private Street", readonly=False, related_sudo=False)
+    private_street2 = fields.Char(related='address_home_id.street2', string="Private Street2", readonly=False, related_sudo=False)
+    private_city = fields.Char(related='address_home_id.city', string="Private City", readonly=False, related_sudo=False)
+    private_state_id = fields.Many2one(
+        related='address_home_id.state_id', string="Private State", readonly=False, related_sudo=False,
+        domain="[('country_id', '=?', private_country_id)]")
+    private_zip = fields.Char(related='address_home_id.zip', readonly=False, string="Private Zip", related_sudo=False)
+    private_country_id = fields.Many2one(related='address_home_id.country_id', string="Private Country", readonly=False, related_sudo=False)
     is_address_home_a_company = fields.Boolean(related='employee_id.is_address_home_a_company', readonly=False, related_sudo=False)
     private_email = fields.Char(related='address_home_id.email', string="Private Email", readonly=False)
     km_home_work = fields.Integer(related='employee_id.km_home_work', readonly=False, related_sudo=False)
@@ -78,6 +86,7 @@ class User(models.Model):
             'active',
             'child_ids',
             'employee_id',
+            'address_home_id',
             'employee_ids',
             'employee_parent_id',
             'hr_presence_state',
@@ -88,7 +97,12 @@ class User(models.Model):
 
         hr_writable_fields = [
             'additional_note',
-            'address_home_id',
+            'private_street',
+            'private_street2',
+            'private_city',
+            'private_state_id',
+            'private_zip',
+            'private_country_id',
             'address_id',
             'barcode',
             'birthday',
